File: runtime/src/main/java/com/huskerdev/nativekt/foreign/ForeignUtils.java
package com.huskerdev.nativekt.foreign;

import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;

@SuppressWarnings("unused")
public class ForeignUtils {

    public static ValueLayout C_CHAR = ValueLayout.JAVA_CHAR;
    public static ValueLayout C_BOOLEAN = ValueLayout.JAVA_BOOLEAN;
    public static ValueLayout C_SHORT = ValueLayout.JAVA_SHORT;
    public static ValueLayout C_INT = ValueLayout.JAVA_INT;
    public static ValueLayout C_LONG = ValueLayout.JAVA_LONG;
    public static ValueLayout C_FLOAT = ValueLayout.JAVA_FLOAT;
    public static ValueLayout C_DOUBLE = ValueLayout.JAVA_DOUBLE;
    public static ValueLayout C_ADDRESS = ValueLayout.ADDRESS;

    public static MemorySegment heapStr(String str) {
        return MemorySegment.ofArray(str.getBytes());
    }

    public static MethodHandle lookup(String name, boolean isCritical, ValueLayout retType, ValueLayout... argTypes) {
        FunctionDescriptor function = retType == null ?
                FunctionDescriptor.ofVoid(argTypes) :
                FunctionDescriptor.of(retType, argTypes);

        MemorySegment address = SymbolLookup.loaderLookup().findOrThrow(name);

        if(isCritical)
            return Linker.nativeLinker().downcallHandle(address, function, Linker.Option.critical(true));
        else
            return Linker.nativeLinker().downcallHandle(address, function);
    }
}

File: modules/runtime/src/jvmMain/kotlin/com/huskerdev/nativekt/jvm/foreign/ForeignUtils.kt
@file:Suppress("unused", "JAVA_MODULE_DOES_NOT_EXPORT_PACKAGE")

package com.huskerdev.nativekt.jvm.foreign

import java.lang.foreign.*
import java.lang.foreign.ValueLayout.*
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.reflect.Method


private val linker = Linker.nativeLinker()

fun lookup(
    lookup: SymbolLookup,
    name: String,
    isCritical: Boolean,
    retType: MemoryLayout?,
    vararg argTypes: MemoryLayout
): MethodHandle {
    val address = lookup.find(name).orElseThrow()

    val function = if (retType == null)
        FunctionDescriptor.ofVoid(*argTypes)
    else FunctionDescriptor.of(retType, *argTypes)

    return if (isCritical)
        linker.downcallHandle(address, function, Linker.Option.critical(true))
    else linker.downcallHandle(address, function)
}

fun upcall(target: Any): MemorySegment {
    val lookup = MethodHandles.privateLookupIn(target.javaClass, MethodHandles.lookup())
    val method = target.javaClass.declaredMethods.first { !it.isSynthetic }
    val methodHandle = lookup.unreflect(method).bindTo(target)
    val descriptor = descriptorFor(method)
    return linker.upcallStub(methodHandle, descriptor, Arena.ofAuto())
}

private fun descriptorFor(method: Method): FunctionDescriptor {
    val args = method.parameterTypes.map(::toLayout)
    val ret = method.returnType
    return if (ret == Void.TYPE)
        FunctionDescriptor.ofVoid(*args.toTypedArray())
    else FunctionDescriptor.of(toLayout(ret), *args.toTypedArray())
}

private fun toLayout(type: Class<*>): MemoryLayout = when (type) {
    MemorySegment::class.java -> ADDRESS
    Byte::class.javaPrimitiveType -> JAVA_BYTE
    Short::class.javaPrimitiveType -> JAVA_SHORT
    Int::class.javaPrimitiveType -> JAVA_INT
    Long::class.javaPrimitiveType -> JAVA_LONG
    Float::class.javaPrimitiveType -> JAVA_FLOAT
    Double::class.javaPrimitiveType -> JAVA_DOUBLE
    Char::class.javaPrimitiveType -> JAVA_CHAR
    Boolean::class.javaPrimitiveType -> JAVA_BOOLEAN
    else -> throw IllegalArgumentException("Unsupported type: ${type.name}")
}